Admin users can mark CMS blocks and CMS elements as favorites

JoshuaBehrens opened the pull request:
### 1. Why is this change necessary?

When you get a lot CMS features from apps but don't use all of them you have to live with it. There is no way to get yourself a structure. Imagine you have someone writing blog entries. They unlikely need to use product listing filter. They still see and stumble upon it. Or you have a content editor that often likes to use a certain block for image and text but forgets which one to use after a week of holidays.
If you could just mark your own favorites and have a useful quick selection of your finest toolings.

See it in action

![PR-cms-fav-1](https://user-images.githubusercontent.com/1133593/179126462-e797a25c-4add-48c9-806f-a41796aee421.gif)

![PR-cms-fav-2](https://user-images.githubusercontent.com/1133593/179126479-d180cc7b-3529-41db-bf0b-103705cca4e0.gif)


### 2. What does this change do, exactly?
Technically a lot:

* We have two new technical settings on users that are similar to sales channel favorites
* Control the chevron direction of a collapsable
* Have new buttons placed to control and display cms favorites via new JS services
* Have display groups to highlight marked favorites


### 3. Describe each step to reproduce the issue or behaviour.

1. Buy and install a feature rich theme in the community store
2. Buy and install an app in the community store that adds bootstrap structural features
3. Buy and install an app in the community store that adds new types of CMS pages and thus adds new CMS blocks
4. Have like 15 block categories and a lot to scroll through when choosing a CMS element
5. Only use like 2 blocks or elements of each provider
6. Search the CMS elements or blocks every time again


### 4. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/workflow/2020-08-03-implement-New-Changelog.md) with all necessary information about my changes
- [x] I have squashed any insignificant commits
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.


### 5. Shoutout

Thanks to several nice persons at Shopware to support me at making this a solution, that matches the overall product. This includes some persons, whose GitHub names I don't know, and in any case @marcelbrode @marie-reckendrees @jleifeld @jkrzefski @itsdenisehere @GitEvil  🎉

taltholtmann commented:
Heyy again :-D. 
We would also need tests for this feature. We have a 100% jest test policy and also e2e tests are required. Otherwise, merging is not possible. Could you add the required tests?

marcelbrode commented on src/Administration/Resources/app/administration/src/module/sw-cms/component/sw-cms-sidebar/index.js:
```diff
@@ -132,9 +133,29 @@ Component.register('sw-cms-sidebar', {
+    created() {
+        if (this.blockTypes.some(blockName => this.cmsBlockFavorites.isFavorite(blockName))) {
+            this.currentBlockCategory = 'favorite';
+        }
```
For the sake of extensibility, please do not put content into "created" besides a single method call `createdComponent();` and proceed with your code in that method.

marcelbrode commented on src/Administration/Resources/app/administration/src/module/sw-cms/component/sw-cms-sidebar/sw-cms-sidebar.scss:
```diff
@@ -88,6 +88,20 @@
     }
 }
 
+.sw-cms-sidebar__block-preview-with-actions {
```
Would be more beautiful if you'd scope that class in `.sw-cms-sidebar`

marcelbrode commented on src/Administration/Resources/app/administration/src/module/sw-cms/component/sw-cms-slot/sw-cms-slot.scss:
```diff
@@ -124,15 +124,29 @@
     }
 }
 
+.sw-cms-slot__modal-container {
```
Again, please scope <3

JoshuaBehrens replied:
Unsure whether I can add something useful. Shall I prepend `.sw-modal`?

<img width="410" alt="image" src="https://user-images.githubusercontent.com/1133593/180645396-a0502cd8-158d-44a5-b5b6-9a26e9952746.png">

marcelbrode replied:
Well, I see, thats hierachically not in sw-cms-slot, but since it's named like that, you should at least put it into ".sw-cms-slot" above, even if you just use `&__model-container` after words, for structure's sake

marcelbrode commented on src/Administration/Resources/app/administration/src/module/sw-cms/snippet/de-DE.json:
```diff
@@ -127,7 +128,8 @@
         "lockedNotification": "Dieses Layout ist gesperrt und kann nicht angepasst werden. Solltest Du dieses Standard-Template, an Deine eigenen Bedürfnisse anpassen möchtest, kannst du es einfach duplizieren und dann das Duplikat anpassen.",
         "headlineEmptyState": "Raum für Kreativität",
-        "claimEmptyState": "Stelle neue Layouts mithilfe von Sektionen zusammen. Wähle eine Sektion und füge einfach per Drag & Drop-Blöcke zu jeder Sektion hinzu und befülle die Blöcke dann mit deinem individuellen Inhalt."
+        "claimEmptyState": "Stelle neue Layouts mithilfe von Sektionen zusammen. Wähle eine Sektion und füge einfach per Drag & Drop-Blöcke zu jeder Sektion hinzu und befülle die Blöcke dann mit deinem individuellen Inhalt.",
```
```suggestion
        "claimEmptyState": "Stelle neue Layouts mithilfe von Sektionen zusammen. Wähle eine Sektion und füge einfach per Drag & Drop Blöcke zu jeder Sektion hinzu und befülle die Blöcke dann mit deinem individuellen Inhalt.",
```

JoshuaBehrens replied:
Hey, that is not my snippet :D

marcelbrode replied:
Boyscouts are always watching :D 
Hope it would be okay anyway <3

marcelbrode commented on src/Administration/Resources/app/administration/src/module/sw-cms/snippet/de-DE.json:
```diff
@@ -127,7 +128,8 @@
         "headlineEmptyState": "Raum für Kreativität",
-        "claimEmptyState": "Stelle neue Layouts mithilfe von Sektionen zusammen. Wähle eine Sektion und füge einfach per Drag & Drop-Blöcke zu jeder Sektion hinzu und befülle die Blöcke dann mit deinem individuellen Inhalt."
+        "claimEmptyState": "Stelle neue Layouts mithilfe von Sektionen zusammen. Wähle eine Sektion und füge einfach per Drag & Drop-Blöcke zu jeder Sektion hinzu und befülle die Blöcke dann mit deinem individuellen Inhalt.",
+        "blockFavoriteEmptyState": "Hier stehen zukünftig deine Favoriten. Klicke auf das Herz an den Blöcken und sammle hier deine am liebsten genutzten Blöcke."
```
@GitEvil

GitEvil replied:
```suggestion
        "blockFavoriteEmptyState": "Hier stehen Deine Favoriten. Favorisiere Blöcke mit einem Klick auf das Herz und finde dadurch schnell Deine meistgenutzten Blöcke wieder."
```

JoshuaBehrens replied:
Do you also have an English suggestion?

marcelbrode replied:
Now he posted it :D

marcelbrode reviewed (changes requested):
Some smaller changes

JoshuaBehrens commented:
> We have a 100% jest test policy and also e2e tests are required

So I can somewhere find tests about these components and just extend them, right?

taltholtmann commented:
> > We have a 100% jest test policy and also e2e tests are required
> 
> So I can somewhere find tests about these components and just extend them, right?

Hopefully, but not necessarily. We introduced the policy for all new code. So, if there are no tests for an old component, you only need to test the methods you created/extended. And all the new files you created need complete test coverage.

GitEvil commented on src/Administration/Resources/app/administration/src/module/sw-cms/snippet/en-GB.json:
```diff
@@ -127,7 +128,8 @@
         "headlineEmptyState": "Space for creativity",
-        "claimEmptyState": "Create new layouts by choosing a section. Start by choosing a section type, drag & drop blocks into any section and fill in your content."
+        "claimEmptyState": "Create new layouts by choosing a section. Start by choosing a section type, drag & drop blocks into any section and fill in your content.",
+        "blockFavoriteEmptyState": "Click on the heart next to the blocks and start collecting favorites to use them from here right away."
```
```suggestion
        "blockFavoriteEmptyState": "This is where you'll find your favourite blocks. Click the heart-shaped symbol next to a block to mark them as favourites and conveniently collect them in one place."
```

JoshuaBehrens committed:
Add e2e test

JoshuaBehrens committed:
Change chevron direction for sw-sidebar-collapse

JoshuaBehrens committed:
Extract base class for user configuration handling

JoshuaBehrens committed:
Add service to handle CMS element favorites

JoshuaBehrens committed:
Add service to handle CMS block favorites

JoshuaBehrens committed:
Fix unrelated snippet

JoshuaBehrens committed:
Add UI for handling CMS block favorites

JoshuaBehrens commented:
I rebased, fixed the tests I see that I broke them, amend the snippet hints from @GitEvil and @marcelbrode and I wrote an e2e test, created an issue on e2e testing docs https://github.com/shopware/docs/issues/642

Handing over to you @taltholtmann  :)

taltholtmann commented:
Hi there, 
in the demo with the QA we found some smaller issues today.
- in the modal, there should be more padding between the favourites and the other elements
- please replace "favorites" with "favourites" in the snipptes :pray: 

![padding](https://user-images.githubusercontent.com/20816234/181446042-7535a568-925f-47b9-9196-7d4b472f5769.png)

- Edited after communication with internal Stakeholders

taltholtmann commented:
Heyy @JoshuaBehrens 
After communication with our internal Stakeholders,  I have edited the above post.

JoshuaBehrens commented:
Sure @taltholtmann , got you covered :)

JoshuaBehrens committed:
Amend translations

JoshuaBehrens committed:
Add changelog

JoshuaBehrens committed:
Add UI for handling CMS element favorites

JoshuaBehrens commented:
There we go. I rebased, fixed the missing padding that I had added before but broke with a change that was asked here and I amend the displayed texts.

JoshuaBehrens commented:
Why do you create more and more conflicts for me. You could've just merged it :(

NiklasLimberg commented:
Thank you for your contribution 💙

JoshuaBehrens commented:
Thank you for your extended efforts in pushing it through :)
